File: code/backend/java/core/src/main/java/games/paths/core/persistence/match/RegistryStoreAdapter.java
package games.paths.core.persistence.match;

import games.paths.core.entity.match.GamingStateRegistryEntity;
import games.paths.core.entity.match.LogEventsEntity;
import games.paths.core.port.match.RegistryStorePort;
import games.paths.core.repository.match.GamingStateRegistryRepository;
import games.paths.core.repository.match.LogEventsRepository;

import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * RegistryStoreAdapter - JPA adapter for {@link RegistryStorePort}, the single writer and
 * reader of {@code gaming_state_registry}. Step 36; Step 36.1 made a key able to own several
 * rows, one per value of its set.
 */
@Repository
public class RegistryStoreAdapter implements RegistryStorePort {

    private final GamingStateRegistryRepository registryRepository;
    private final LogEventsRepository logEventsRepository;
    private final games.paths.core.repository.match.GamingMatchRepository matchRepository;

    public RegistryStoreAdapter(GamingStateRegistryRepository registryRepository,
                                LogEventsRepository logEventsRepository,
                                games.paths.core.repository.match.GamingMatchRepository matchRepository) {
        this.registryRepository = registryRepository;
        this.logEventsRepository = logEventsRepository;
        this.matchRepository = matchRepository;
    }

    @Override
    @Transactional(readOnly = true)
    public java.util.Optional<long[]> findMatchAndStoryIdByUuid(String matchUuid) {
        if (matchUuid == null || matchUuid.isBlank()) {
            return java.util.Optional.empty();
        }
        return matchRepository.findByUuid(matchUuid)
                .filter(m -> m.getId() != null && m.getIdStory() != null)
                .map(m -> new long[]{m.getId(), m.getIdStory()});
    }

    @Override
    @Transactional(readOnly = true)
    public List<RegistryRow> findByMatch(long idMatch) {
        List<RegistryRow> out = new ArrayList<>();
        for (GamingStateRegistryEntity r : registryRepository.findByIdMatch(idMatch)) {
            out.add(toRow(r));
        }
        return out;
    }

    @Override
    @Transactional(readOnly = true)
    public List<RegistryRow> findByMatchAndKey(long idMatch, String key) {
        if (key == null) {
            return List.of();
        }
        List<RegistryRow> out = new ArrayList<>();
        for (GamingStateRegistryEntity r : registryRepository.findByIdMatchAndKey(idMatch, key)) {
            out.add(toRow(r));
        }
        return out;
    }

    @Override
    @Transactional
    public void upsert(long idMatch, String key, String stringValue, Integer intValue,
                       Long idCharacter, Long idEvent, Long idChoice, Integer clock) {
        List<GamingStateRegistryEntity> rows = registryRepository.findByIdMatchAndKey(idMatch, key);
        GamingStateRegistryEntity row = rows.isEmpty() ? null : rows.get(0);
        if (row == null) {
            row = new GamingStateRegistryEntity();
            row.setId(nextId(idMatch));
            row.setIdMatch(idMatch);
            row.setKey(key);
            row.setMultiValue(0);
        }
        row.setStringValue(stringValue);
        row.setIntValue(intValue);
        row.setIdCharacter(idCharacter);
        row.setIdEvent(idEvent);
        row.setIdChoice(idChoice);
        row.setClock(clock);
        registryRepository.save(row);
    }

    @Override
    @Transactional
    public void insertValue(long idMatch, String key, String stringValue, Integer intValue,
                            Long idCharacter, Long idEvent, Long idChoice, Integer clock) {
        GamingStateRegistryEntity row = new GamingStateRegistryEntity();
        row.setId(nextId(idMatch));
        row.setIdMatch(idMatch);
        row.setKey(key);
        row.setMultiValue(1);
        row.setStringValue(stringValue);
        row.setIntValue(intValue);
        row.setIdCharacter(idCharacter);
        row.setIdEvent(idEvent);
        row.setIdChoice(idChoice);
        row.setClock(clock);
        registryRepository.save(row);
    }

    @Override
    @Transactional
    public void deleteValue(long idMatch, String key, String stringValue, Integer intValue) {
        for (GamingStateRegistryEntity r : registryRepository.findByIdMatchAndKey(idMatch, key)) {
            if (Objects.equals(r.getStringValue(), stringValue)
                    && Objects.equals(r.getIntValue(), intValue)) {
                registryRepository.delete(r);
                return;
            }
        }
    }

    @Override
    @Transactional
    public void insertAll(long idMatch, List<RegistryRow> rows) {
        if (rows == null || rows.isEmpty()) {
            return;
        }
        List<GamingStateRegistryEntity> entities = new ArrayList<>();
        long nextId = 1L;
        for (RegistryRow r : rows) {
            GamingStateRegistryEntity e = new GamingStateRegistryEntity();
            e.setId(nextId++);
            e.setIdMatch(idMatch);
            e.setKey(r.key());
            e.setStringValue(r.stringValue());
            e.setIntValue(r.intValue());
            e.setMultiValue(r.isMulti() ? 1 : 0);
            entities.add(e);
        }
        registryRepository.saveAll(entities);
    }

    @Override
    @Transactional
    public void deleteByMatchIdIn(List<Long> matchIds) {
        if (matchIds == null || matchIds.isEmpty()) {
            return;
        }
        registryRepository.deleteByMatchIdIn(matchIds);
    }

    @Override
    @Transactional
    public void logChange(long idMatch, Long idCharacter, Long idEvent, Long idChoice,
                          Integer clock, String message) {
        LogEventsEntity e = new LogEventsEntity();
        e.setId(logEventsRepository.findMaxId() + 1);
        e.setIdMatch(idMatch);
        e.setIdCharacterMatch(idCharacter);
        e.setIdEvent(idEvent);
        e.setIdChoise(idChoice);
        e.setClock(clock);
        e.setLogMessage(message);
        logEventsRepository.save(e);
    }

    private long nextId(long idMatch) {
        return registryRepository.findByIdMatch(idMatch).stream()
                .map(GamingStateRegistryEntity::getId)
                .filter(Objects::nonNull)
                .mapToLong(Long::longValue).max().orElse(0L) + 1L;
    }

    private static RegistryRow toRow(GamingStateRegistryEntity r) {
        return new RegistryRow(r.getId(), r.getUuid(), r.getKey(), r.getStringValue(),
                r.getIntValue(), r.getIdCharacter(), r.getIdEvent(), r.getIdChoice(), r.getClock(),
                r.getMultiValue());
    }
}

File: code/backend/python/app/adapters/persistence/match/registry_store_adapter.py
"""Step 36 — SQLAlchemy adapter for the registry store port.

The single reader and writer of gaming_state_registry, and the writer of the log_events audit
row every registry change leaves behind.
"""
import uuid as uuid_lib
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from app.adapters.persistence.match.models import (GamingMatchEntity,
                                                    GamingStateRegistryEntity, LogEventsEntity)
from app.core.ports.match.registry_ports import RegistryStorePort


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def _to_row(r: GamingStateRegistryEntity) -> Dict[str, Any]:
    return {
        "id": r.id,
        "uuid": r.uuid,
        "key": r.key,
        "string_value": r.string_value,
        "int_value": r.int_value,
        "id_character": r.id_character,
        "id_event": r.id_event,
        "id_choice": r.id_choice,
        "clock": r.clock,
        "multi_value": r.multi_value,
    }


class RegistryStoreAdapter(RegistryStorePort):

    def __init__(self, session_factory) -> None:
        self.session_factory = session_factory

    def find_match_and_story_id_by_uuid(self, match_uuid: str) -> Optional[tuple]:
        if not match_uuid or not str(match_uuid).strip():
            return None
        with self.session_factory() as session:
            m = (session.query(GamingMatchEntity)
                 .filter(GamingMatchEntity.uuid == match_uuid).first())
            if m is None or m.id is None or m.id_story is None:
                return None
            return (m.id, m.id_story)

    def find_by_match(self, id_match: int) -> List[Dict[str, Any]]:
        with self.session_factory() as session:
            rows = (session.query(GamingStateRegistryEntity)
                    .filter(GamingStateRegistryEntity.id_match == id_match).all())
            return [_to_row(r) for r in rows]

    def find_by_match_and_key(self, id_match: int, key: str) -> List[Dict[str, Any]]:
        if key is None:
            return []
        with self.session_factory() as session:
            rows = (session.query(GamingStateRegistryEntity)
                    .filter(GamingStateRegistryEntity.id_match == id_match,
                            GamingStateRegistryEntity.key == key).all())
            return [_to_row(r) for r in rows]

    def upsert(self, id_match: int, key: str, string_value: Optional[str],
               int_value: Optional[int], id_character: Optional[int],
               id_event: Optional[int], id_choice: Optional[int],
               clock: Optional[int]) -> None:
        with self.session_factory() as session:
            row = (session.query(GamingStateRegistryEntity)
                   .filter(GamingStateRegistryEntity.id_match == id_match,
                           GamingStateRegistryEntity.key == key).first())
            now = _now_iso()
            if row is None:
                row = GamingStateRegistryEntity(
                    id=self._next_id(session, id_match), id_match=id_match,
                    uuid=str(uuid_lib.uuid4()), key=key, multi_value=0,
                    ts_insert=now, ts_update=now)
                session.add(row)
            row.string_value = string_value
            row.int_value = int_value
            row.id_character = id_character
            row.id_event = id_event
            row.id_choice = id_choice
            row.clock = clock
            row.ts_update = now
            session.commit()

    def insert_value(self, id_match: int, key: str, string_value: Optional[str],
                     int_value: Optional[int], id_character: Optional[int],
                     id_event: Optional[int], id_choice: Optional[int],
                     clock: Optional[int]) -> None:
        with self.session_factory() as session:
            now = _now_iso()
            session.add(GamingStateRegistryEntity(
                id=self._next_id(session, id_match), id_match=id_match,
                uuid=str(uuid_lib.uuid4()), key=key, multi_value=1,
                string_value=string_value, int_value=int_value,
                id_character=id_character, id_event=id_event, id_choice=id_choice,
                clock=clock, ts_insert=now, ts_update=now))
            session.commit()

    def delete_value(self, id_match: int, key: str, string_value: Optional[str],
                     int_value: Optional[int]) -> None:
        with self.session_factory() as session:
            row = (session.query(GamingStateRegistryEntity)
                   .filter(GamingStateRegistryEntity.id_match == id_match,
                           GamingStateRegistryEntity.key == key,
                           GamingStateRegistryEntity.string_value == string_value,
                           GamingStateRegistryEntity.int_value == int_value).first())
            if row is not None:
                session.delete(row)
                session.commit()

    def insert_all(self, id_match: int, rows: List[Dict[str, Any]]) -> None:
        if not rows:
            return
        with self.session_factory() as session:
            now = _now_iso()
            for index, r in enumerate(rows, start=1):
                session.add(GamingStateRegistryEntity(
                    id=index, id_match=id_match, uuid=str(uuid_lib.uuid4()),
                    key=r.get("key"), string_value=r.get("string_value"),
                    int_value=r.get("int_value"), multi_value=r.get("multi_value") or 0,
                    ts_insert=now, ts_update=now))
            session.commit()

    def delete_by_match_ids(self, match_ids: List[int]) -> None:
        if not match_ids:
            return
        with self.session_factory() as session:
            (session.query(GamingStateRegistryEntity)
             .filter(GamingStateRegistryEntity.id_match.in_(match_ids))
             .delete(synchronize_session=False))
            session.commit()

    def log_change(self, id_match: int, id_character: Optional[int], id_event: Optional[int],
                   id_choice: Optional[int], clock: Optional[int], message: str) -> None:
        with self.session_factory() as session:
            rows = session.query(LogEventsEntity.id).all()
            ids = [r[0] for r in rows if r[0] is not None]
            now = _now_iso()
            session.add(LogEventsEntity(
                id=(max(ids) if ids else 0) + 1, id_match=id_match,
                uuid=str(uuid_lib.uuid4()), id_character_match=id_character,
                timestamp=now, id_event=id_event, id_choise=id_choice,
                log_message=message, clock=clock, ts_insert=now, ts_update=now))
            session.commit()

    def _next_id(self, session, id_match: int) -> int:
        rows = (session.query(GamingStateRegistryEntity.id)
                .filter(GamingStateRegistryEntity.id_match == id_match).all())
        ids = [r[0] for r in rows if r[0] is not None]
        return (max(ids) if ids else 0) + 1
